plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1678,7 +1694,7 @@ private void flushBuffer(boolean finished)
             }
 
             // The multipart upload API only accept the last part to be less than 5MB
-            if (bufferSize == buffer.length || (finished && bufferSize > 0)) {
+            if (bufferSize == partSize || (finished && bufferSize > 0)) {
```

Comment:
Below:
```
                else {
                    this.buffer = new byte[buffer.length];
                    bufferSize = 0;
                }
```

Instead of using `new byte[buffer.length]` you could start with `new byte[0]` and increase to `partSize` after first write